Bottom: 70c6d75ef5e17eb4c1feecccad965b45a1bdb8af
Top:    3ac6ab007adb37525f56b8e4f0f8b21b19fee3e0
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-11-03 23:47:22 +0000

Tag circular contours so they can be drawn quicker


---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index 54c2b3b..76953d1 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -602,10 +602,25 @@ do_hole (const BoxType *b, void *cl)
 {
   struct do_hole_info *info = cl;
   PLINE *curc = (PLINE *) b;
+
   /* Ignore the outer contour - we draw it first explicitly*/
   if (curc->Flags.orient == PLF_DIR) {
     return 0;
   }
+
+  /* If the contour is round, and hidgl_fill_circle would use
+   * less slices than we have vertices to draw it, then call
+   * hidgl_fill_circle to draw this contour.
+   */
+  if (curc->is_round) {
+    double slices = M_PI * 2 * curc->radius /
+                    info->scale / PIXELS_PER_CIRCLINE;
+    if (slices < curc->Count) {
+      hidgl_fill_circle (curc->cx, curc->cy, curc->radius, info->scale);
+      return 1;
+    }
+  }
+
   tesselate_contour (info->tobj, &curc->head, info->vertices);
   return 1;
 }
diff --git a/src/polyarea.h b/src/polyarea.h
index 50dda4a..ce427ad 100644
--- a/src/polyarea.h
+++ b/src/polyarea.h
@@ -99,6 +99,10 @@ struct PLINE
     unsigned int Count;
     double area;
     rtree_t *tree;
+    int is_round;
+    int cx;
+    int cy;
+    int radius;
     struct {
       unsigned int status:3;
       unsigned int orient:1;
diff --git a/src/polygon.c b/src/polygon.c
index a0d0bef..0a86fa4 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -351,6 +351,10 @@ CirclePoly (LocationType x, LocationType y, BDimension radius)
   if ((contour = poly_NewContour (v)) == NULL)
     return NULL;
   frac_circle (contour, x, y, v, 1);
+  contour->is_round = TRUE;
+  contour->cx = x;
+  contour->cy = y;
+  contour->radius = radius;
   return ContourToPoly (contour);
 }
 
diff --git a/src/polygon1.c b/src/polygon1.c
index b932f35..b7f6e6c 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -2510,6 +2510,10 @@ poly_IniContour (PLINE * c)
   c->head.next = c->head.prev = &c->head;
   c->xmin = c->ymin = 0x7fffffff;
   c->xmax = c->ymax = 0x80000000;
+  c->is_round = FALSE;
+  c->cx = 0;
+  c->cy = 0;
+  c->radius = 0;
 }
 
 PLINE *
